code/modules/mob/living/silicon/ai/laws.dm

```diff
@@ -17,10 +17,16 @@
 
 	src.laws_sanity_check()
 	src.laws.show_laws(who)
-	if(!everyone)
-		for(var/mob/living/silicon/robot/R in connected_robots)
-			if(R.lawupdate)
-				R.lawsync()
-				R.show_laws()
-				R.law_change_counter++
 
+/mob/living/silicon/ai/post_lawchange(announce = TRUE)
+	. = ..()
+
+	if(!.)
+		return
+
+	addtimer(CALLBACK(src, .proc/lawsync), 0)
```

Comment:
Why are you make a zero-time call?
There's invoke async if there's sleeping procs there, but is there any?

Reply:
Sometimes `post_lawchange` is called multiple times in a tick so it waits until the next tick.

Reply:
This won't change the fact that it will be called multiple times in a tick. It will just happen at the start of the next.
Why is it called so much, how come the laws change that much?

Reply:
```suggestion
	// lawset modules cause this function to be executed multiple times in a tick, so we wait for the next tick in order to be able to see the entire lawset
	addtimer(CALLBACK(src, .proc/lawsync), 0)
```

Reply:
Would you like this changed before this can be merged?

Reply:
Depends on if you want to do a major refactor or not.
Adding a timer like this won't work, because you'll just be adding several timers.
If you do want to add unique timers you can make a macro and replace all of those, something like this:
`#define LAW_SYNC(source) addtimer(CALLBACK(source, .proc/lawsync), 1, TIMER_UNIQUE)`
Then all the changes will be done a single time next tick.

Reply:
Have you looked at the parent proc yet? It returns before it can run if `world.time` hasn't changed. I personally don't like how it works either but I have nothing planned to require changing it.

Reply:
That's fine. Just make the timer unique then, so we don't spam a ton of timers.